### PR TITLE
Allow to skip absent Hive partition

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 1]
```diff
@@ -146,6 +146,8 @@ Property Name                                      Description
                                                    ignored. This is equivalent to the
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]: > Allow to skip empty Hive partition

Please update the commit message as well

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -160,7 +161,9 @@ public BackgroundHiveSplitLoader(
             Executor executor,
             int loaderConcurrency,
             boolean recursiveDirWalkerEnabled,
+            boolean ignoreAbsentPartitions,
             Optional<ValidWriteIdList> validWriteIds)
+
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: revert

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -122,6 +122,7 @@
     private final DirectoryLister directoryLister;
     private final int loaderConcurrency;
     private final boolean recursiveDirWalkerEnabled;
+    private final boolean ignoreAbsentPartitions;
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: try add this consistently in: fields, ctor params, ctor assignments

i know the existing code is a bit messy from this point of view, but eg keep the new filed always right after `recursiveDirWalkerEnabled `

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 4]
```diff
@@ -130,8 +147,13 @@ private FileStatusIterator(Table table, Path path, FileSystem fileSystem, Direct
         {
             this.path = path;
             this.namenodeStats = namenodeStats;
+            this.fileStatusIterator = list(table, path, fileSystem, directoryLister);
+        }
+
+        private RemoteIterator<LocatedFileStatus> list(Table table, Path path, FileSystem fileSystem, DirectoryLister directoryLister)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: unnecessary change?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 5]
```diff
@@ -51,23 +52,26 @@
     private final DirectoryLister directoryLister;
     private final NamenodeStats namenodeStats;
     private final NestedDirectoryPolicy nestedDirectoryPolicy;
+    private final boolean ignoreAbsentPartitions;
 
-    private Iterator<LocatedFileStatus> remoteIterator = Collections.emptyIterator();
+    private Iterator<LocatedFileStatus> remoteIterator = emptyIterator();
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: It used to use `Collections.emptyIterator()`, now it uses `emptyIterator`. 
This looks like a non intentional change, please revert.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 6]
```diff
@@ -168,6 +190,24 @@ private PrestoException processException(IOException exception)
             }
             return new PrestoException(HIVE_FILESYSTEM_ERROR, "Failed to list directory: " + path, exception);
         }
+
+        private static RemoteIterator<LocatedFileStatus> emptyIterator()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: It looks redundant over `Collections.emptyIterator`. Please remove

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestIgnoreAbsentPartition.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        String tableNameInDatabase = tablesState.get("test_table").getNameInDatabase();
+
+        assertThat(query("SELECT count(*) FROM " + tableNameInDatabase)).containsOnly(row(15));
+
+        String partitionPath = format("/user/hive/warehouse/%s/p_regionkey=9999", tableNameInDatabase);
+        assertFalse(hdfsClient.exist(partitionPath));
+        query(format("CALL hive.system.create_empty_partition('default', '%s', array['p_regionkey'], array['9999'])", tableNameInDatabase));
+        assertTrue(hdfsClient.exist(partitionPath));
+        hdfsClient.delete(partitionPath);
+        assertFalse(hdfsClient.exist(partitionPath));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestIgnoreAbsentPartition.java]: please add a message to assertFalse/True, otherwise the raised exc msg will be little hard to use

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestIgnoreAbsentPartition.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    @Inject
+    private HdfsClient hdfsClient;
+
+    @Override
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return mutableTable(NATION_PARTITIONED_BY_BIGINT_REGIONKEY, "test_table", LOADED);
+    }
+
+    /**
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestIgnoreAbsentPartition.java]:
```suggestion
    /*
```

because the comment is not a javadoc comment

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java (hunk 4 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: Please elaborate. Method is used in the ctor of FileStatusIterator

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: Yes, the commit extracts the method while it was "inline" in the ctor.
It seems like a redundant unrelated change, so please inline it back (unless i am missing something)

Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: From my perspective it improves readability - I can inline back (revert) this change if it's unnecessary

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: OTOH it's a refactor, so would belong to a separate commit.

(I think it's redundant, so i'd prefer to inline, but no strong feeling about this.)

**Comment by findepi** [comment]
@electrum do you want to take a look?

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 9]
```diff
@@ -146,6 +146,10 @@ Property Name                                      Description
...
                                                    ``hive.mapred.supports.subdirectories`` property in Hive.
 
+``hive.ignore-absent-partitions``                  Ignore partitions when the location does not exist rather    ``false``
```
Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive.rst]: Sorry, one more change. Let's clarify this as "file system location"

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 10]
```diff
@@ -130,10 +146,10 @@ private FileStatusIterator(Table table, Path path, FileSystem fileSystem, Direct
         {
             this.path = path;
             this.namenodeStats = namenodeStats;
+
             try {
                 this.fileStatusIterator = directoryLister.list(fileSystem, table, path);
-            }
-            catch (IOException e) {
+            } catch(IOException e) {
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: Revert this formatting change

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 11]
```diff
@@ -173,6 +175,7 @@ public BackgroundHiveSplitLoader(
         this.recursiveDirWalkerEnabled = recursiveDirWalkerEnabled;
         this.executor = executor;
         this.partitions = new ConcurrentLazyQueue<>(partitions);
+        this.ignoreAbsentPartitions = ignoreAbsentPartitions;
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java]: Move to after `recursiveDirWalkerEnabled`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 12]
```diff
@@ -108,10 +111,23 @@ protected LocatedFileStatus computeNext()
     private Iterator<LocatedFileStatus> getLocatedFileStatusRemoteIterator(Path path)
     {
         try (TimeStat.BlockTimer ignored = namenodeStats.getListLocatedStatus().time()) {
+            if (ignoreAbsentPartitions && !exists(path)) {
+                return Collections.emptyIterator();
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: Nit: static import the method since the name is self-descriptive and thus the `Collections` prefix doesn't provide any value

**Comment by wendigo** [comment]
@electrum done: ```SET SESSION hive.ignore_absent_partitions = false```
@findepi please review

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 13]
```diff
@@ -146,6 +146,10 @@ Property Name                                      Description
                                                    ignored. This is equivalent to the
                                                    ``hive.mapred.supports.subdirectories`` property in Hive.
 
+``hive.ignore-absent-partitions``                  Ignore partitions when the file system location does not     ``false``
+                                                   exist rather than failing the query. This skips data that
```
Comment (ankitdixit) [inline presto-docs/src/main/sphinx/connector/hive.rst]: nit, please feel free to ignore: 
* add a "," after exist. 
* line "This skips data that may be expected to be part of the table" does not seem to be adding more context.

Comment (wendigo) [reply presto-docs/src/main/sphinx/connector/hive.rst]: @ankitdixit it's exactly as @electrum requested :)

**Comment by wendigo** [comment]
@electrum I've tried to move tests from product to smoke but ```TestHiveIntegrationSmokeTest``` uses ```FileHiveMetastore``` which in turn does not cache partitions list (instead it reads it every time from disk) so deleting partition on disk doesn't end in ```BackgroundHiveSplitLoader``` trying to read non-existing partition.

**Comment by electrum** [comment]
What happens for unpartitioned tables?

**Comment by wendigo** [comment]
Good question @electrum. It will kick-in and I guess it's not expected behaviour.

**Comment by wendigo** [comment]
@electrum https://github.com/prestosql/presto/pull/2555/commits/631d94f7d244311b2a6320cb5a2553be7ec3fef7

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java [hunk 14]
```diff
@@ -130,6 +146,7 @@ private FileStatusIterator(Table table, Path path, FileSystem fileSystem, Direct
         {
             this.path = path;
             this.namenodeStats = namenodeStats;
+
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveFileIterator.java]: revert

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    private HdfsClient hdfsClient;
+
+    @Override
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return mutableTable(NATION_PARTITIONED_BY_BIGINT_REGIONKEY, "test_table", LOADED);
+    }
+
+    @Test
+    public void testIgnoreAbsentPartitionsWithSessionProperties()
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: remove `WithSessionProperties` part because it doesn't add more information

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java [hunk 16]
```diff
@@ -207,6 +207,8 @@ public ConnectorSplitSource getSplits(
 
         Iterable<HivePartitionMetadata> hivePartitions = getPartitionMetadata(session, metastore, table, tableName, partitions, bucketHandle.map(HiveBucketHandle::toTableBucketProperty));
 
+        boolean ignoreAbsentPartitions = !hiveTable.getPartitionColumns().isEmpty() && HiveSessionProperties.isIgnoreAbsentPartitions(session);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java]: i'd static import the `isIgnoreAbsentPartitions` method and inline the variable

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(query("SELECT count(*) FROM " + tableNameInDatabase)).containsOnly(row(15));
+    }
+
+    @Test
+    public void testShouldThrowErrorOnUnpartitionedTableMissingData()
+    {
+        String tableName = "unpartitioned_absent_table_data";
+        String tablePath = "/user/hive/warehouse/" + tableName;
+
+        assertThat(query(format("CREATE TABLE IF NOT EXISTS %s (col INT)", tableName))).containsOnly(row(0));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: remove `IF NOT EXISTS`
if you want the test to be repeatable (even if fails), add `"DROP TABLE " + tableName` before the `CREATE`.
This ensures this is _our_ test table, with expected data & schema, not something else.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(query(format("INSERT INTO %s (col) VALUES (1), (2), (3)", tableName))).containsOnly(row(3));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: - just `VALUES 1, 2, 3` would do the same
- you can combine this with create:
  ```
  CREATE TABLE unpartitioned_absent_table_data AS VALUES 1, 2, 3
  ```

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(query("SELECT COUNT(1) FROM " + tableName)).containsOnly(row(3));
+
+        assertTrue(hdfsClient.exist(tablePath));
+        hdfsClient.delete(tablePath);
+
+        query("SET SESSION hive.ignore_absent_partitions = false");
+        assertThat(() -> query("SELECT COUNT(1) FROM " + tableName)).failsWithMessage("Partition location does not exist");
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: Since the table is not partitioned, the message should say "Table location".
But i understand that's pre-existing, let's not change this now.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+        query("SET SESSION hive.ignore_absent_partitions = true");
+        assertThat(() -> query("SELECT COUNT(1) FROM " + tableName)).failsWithMessage("Partition location does not exist");
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: Here & above, per our conventions:
- lowercase `count` (uppercase keywords from the grammar, generally lowercase everything else)
- `count(*)` is "more typical" than `count(1)`

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+        assertThat(query("DROP TABLE IF EXISTS " + tableName));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveIgnoreAbsentPartitions.java]: remove `IF EXISTS`, as it must exist at this point

you don't need to `assert` here. It's just cleanup, not part of the test.

**Comment by wendigo** [comment]
Squashed @findepi

**Comment by wendigo** [comment]
Failing test is ```io.prestosql.operator.TestTopNRowNumberOperator```

**Comment by findepi** [comment]
CI failed -- https://github.com/prestosql/presto/issues/2616. Retriggered.

**Comment by findepi** [comment]
Merged, thanks!
